ttt.c

```diff
@@ -30,50 +38,94 @@ void draw_board(const char *t)
+    for (int i = 0; i < BOARD_SIZE; i++)
         printf("---");
     printf("\n");
     printf("    ");
-    for (int i = 0; i < N; i++)
+    for (int i = 0; i < BOARD_SIZE; i++)
         printf(" %2c", 'A' + i);
     printf("\n");
 }
 
+char check_row_col(const char *t, int flag)
+{
+    int i_shift[2], j_shift[2];
+    if (flag == ROW) {
+        memcpy(i_shift, no_step, sizeof(i_shift));
+        memcpy(j_shift, step_forward, sizeof(j_shift));
+    } else if (flag == COL) {
+        memcpy(i_shift, step_forward, sizeof(i_shift));
+        memcpy(j_shift, no_step, sizeof(j_shift));
+    }
+    for (int i = 0; i < BOARD_SIZE - i_shift[1]; i++) {
+        for (int j = 0; j < BOARD_SIZE - j_shift[1]; j++) {
+            if (t[get_index(i, j)] != ' ' &&
+                t[get_index(i, j)] ==
+                    t[get_index(i + i_shift[0], j + j_shift[0])] &&
+                t[get_index(i + i_shift[0], j + j_shift[0])] ==
+                    t[get_index(i + i_shift[1], j + j_shift[1])]) {
+                return t[get_index(i, j)];
+            }
+        }
+    }
+    return ' ';
+}
+
+char check_diagonal(const char *t, int flag)
+{
+    int i_shift[2], j_shift[2];
+    if (flag == MAIN) {
+        memcpy(i_shift, step_forward, sizeof(i_shift));
+        memcpy(j_shift, step_forward, sizeof(j_shift));
+    } else if (flag == BACK) {
+        memcpy(i_shift, step_backward, sizeof(i_shift));
+        memcpy(j_shift, step_forward, sizeof(j_shift));
+    }
+    for (int i = 0; i < BOARD_SIZE; i++) {
+        for (int j = 0; j < BOARD_SIZE; j++) {
+            if (i + i_shift[1] < 0 || i + i_shift[1] >= BOARD_SIZE ||
+                j + j_shift[1] < 0 || j + j_shift[1] >= BOARD_SIZE)
+                continue;
+            if (t[get_index(i, j)] != ' ' &&
+                t[get_index(i, j)] ==
+                    t[get_index(i + i_shift[0], j + j_shift[0])] &&
+                t[get_index(i + i_shift[0], j + j_shift[0])] ==
+                    t[get_index(i + i_shift[1], j + j_shift[1])]) {
+                return t[get_index(i, j)];
+            }
+        }
+    }
+    return ' ';
+}
+
 char check_win(char *t)
 {
-    if (t[0] != ' ' && t[0] == t[1] && t[1] == t[2])
-        return t[0];
-    if (t[3] != ' ' && t[3] == t[4] && t[4] == t[5])
-        return t[3];
-    if (t[6] != ' ' && t[6] == t[7] && t[7] == t[8])
-        return t[6];
-
-    if (t[0] != ' ' && t[0] == t[3] && t[3] == t[6])
-        return t[0];
-    if (t[1] != ' ' && t[1] == t[4] && t[4] == t[7])
-        return t[1];
-    if (t[2] != ' ' && t[2] == t[5] && t[5] == t[8])
-        return t[2];
-
-    if (t[0] != ' ' && t[0] == t[4] && t[4] == t[8])
-        return t[0];
-    if (t[2] != ' ' && t[2] == t[4] && t[4] == t[6])
-        return t[2];
-
-    for (int i = 0; i < 9; i++)
+    char r = check_row_col(t, ROW);
+    if (r != ' ')
+        return r;
+    char c = check_row_col(t, COL);
+    if (c != ' ')
+        return c;
+    char md = check_diagonal(t, MAIN);
```

Comment:
The parameters of `check_diagonal`, namely `MAIN` and `BACK`, are ambiguous, making them hard to distinguish. Can you improve?

Reply:
Refering https://en.wikipedia.org/wiki/Main_diagonal
Would `PRIMARY` and `SECONDARY` be a greater name?

Reply:
> Would PRIMARY and SECONDARY be a greater name?

It sounds good.